# Tutorial

## Summary of a typical `scbs` workflow

1. use `scbs prepare` to store sc-methylation data in an efficient format
2. use `scbs scan` to discover methylation-variable regions in the genome, or provide your own regions of interest
3. use `scbs matrix` to receive a matrix analogous to the count matrix in scRNA-seq
4. use the matrix for downstream analysis such as dimensionality reduction and clustering


### What you will need

`scbs` assumes that you
- used single-cell bisulfite-sequencing to generate FASTQ-files of bisulfite-converted reads
- mapped the reads with a methylation-aware program
- extracted context-dependent methylation values (typically CpG) from the alignments

The last two steps can be achieved with bulk bisulfite-sequencing tools such as [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/) and the included `bismark_methylation_extractor` script.
Afterwards, you should be left with a simple tabular file for each cell.
Each file contains the methylation status of thousands of methylation sites (typically CpG-sites) in the whole genome.
Here are the first five lines of an example file, which is a `.cov`-file generated by Bismark:

```
1       100021800       100021800       100.000000      1       0
1       100023064       100023064       0.000000        0       1
1       100026855       100026855       100.000000      1       0
1       100027218       100027218       100.000000      2       0
1       100051671       100051671       0.000000        0       2
```

You need one of these files per cell.
The columns denote the chromosome name, the start and end coordinates of the methylation site (identical in this case), the observed percentage of methylation (typically 0% or 100% in single-cell data), the number of reads that are methylated at that site, and the number of unmethylated reads.
If you did not use Bismark and your files have a slightly different format, don't worry.
We support a range of different input formats and you can even define your own custom format.


### 1. Using `scbs prepare`

to do