import click
import numba
from click import echo, style
from datetime import datetime, timedelta
from scbs.scbs import _get_filepath, profile, prepare, smooth, matrix, scan
from click_help_colors import HelpColorsGroup


class Timer(object):
    def __init__(self, label="run", fmt="%a %b %d %H:%M:%S %Y"):
        self.label = style(label, bold=True)
        self.fmt = fmt
        self.begin_time = datetime.now()
        echo(f"\nStarted {self.label} on {self.begin_time.strftime(self.fmt)}.")
        return

    def stop(self):
        end_time = datetime.now()
        runtime = timedelta(seconds=(end_time - self.begin_time).seconds)
        echo(
            f"\nFinished {self.label} on {end_time.strftime(self.fmt)}. "
            f"Total runtime was {runtime} (hour:min:s)."
        )
        return


def _print_kwargs(kwargs):
    echo("\nCommand line arguments:")
    for arg, value in kwargs.items():
        if value is not None:
            value_fmt = style(str(_get_filepath(value)), fg="blue")
            echo(f"{arg: >15}: {value_fmt}")
    echo()


def _set_n_threads(ctx, param, value):
    """
    Set the number of CPU threads for numba.
    Arguments come straight from click option.
    """
    if value == -1:
        return numba.config.NUMBA_NUM_THREADS
    elif value == 0:
        return 1
    else:
        return value


# command group
@click.group(
    cls=HelpColorsGroup,
    help_headers_color="bright_white",
    help_options_color="green",
    help=f"""
        Below you find a list of all available commands.
        To find out what they do and how to use them, check
        their help like this:

        {style("scbs profile --help", fg="blue")}

        To use stdin or stdout, use the character
        {style("-", fg="blue")} instead of a file path.
        """,
)
@click.version_option()
def cli():
    pass


# prepare command
@cli.command(
    name="prepare",
    help=f"""
    Gathers single cell methylation data from multiple input files
    (one per cell) and creates a sparse matrix (position x cell) in CSR
    format for each chromosome. Methylated sites are represented by a 1,
    unmethylated sites are -1, missing values and other bases are 0.

    {style("INPUT_FILES", fg="green")} are single cell
    methylation files, for example '.cov'-files generated by Bismark.

    {style("DATA_DIR", fg="green")} is the output directory
    where the methylation data will be stored.

    Note: If you have many cells and encounter a "too many open files"-
    error, you need to increase the open file limit with e.g.
    'ulimit -n 9999'.
    """,
    short_help="First step: Collect and store sc-methylation data for quick access",
    no_args_is_help=True,
)
@click.argument("input-files", type=click.File("rb"), nargs=-1)
@click.argument(
    "data-dir",
    type=click.Path(dir_okay=True, file_okay=False, writable=True),
)
@click.option(
    "--input-format",
    default="bismark",
    help="""
    Specify the format of the input files. When using Bismark's '.cov' format,
    type --format bismark. When using using methylpy's 'allc' format, type --input-format allc.
    
    \b
    If none of the above apply, specify a costum format enclosed by quotation marks:
    The positions should be separate with ':' and denote:
    \b
    1. No. (1-indexed) of column with chromosome name
    2. No. (1-indexed) of column with genomic position
    3. No. (1-indexed) of column with methylated counts
    4. No. (1-indexed)of column with unmethylated counts(m) / coverage(c). Depending on which of the two is used, specify with 'm' or 'c'.
    5. Type of separation, eg '\\t' for tab-separation or ',' for comma-separation
    6. 0 if the file has a header, 1 if the file does not have a header
    \b
    Example: --input-format '1:2:3:4m:\\t:1'

    If nothing is specified, the default format is bismark (--input-format '1:2:5:6c:\\t:0'); --input-format 'allc' correspods to --input-format '1:2:5:6m:\\t:1'""",
)
@click.option(
    "--header",
    is_flag=True,
    help="Use this when input files have a column header [default: off, not "
    "required when using Bismark files]",
)
def prepare_cli(**kwargs):
    timer = Timer(label="prepare")
    _print_kwargs(kwargs)
    prepare(**kwargs)
    timer.stop()


# profile command
@cli.command(
    name="profile",
    help=f"""
    From single cell methylation or NOMe-seq data,
    calculates the average methylation profile of a set of
    genomic regions. Useful for plotting and visually comparing
    methylation between groups of regions or cells.

    {style("REGIONS", fg="green")} is an alphabetically sorted (!) .bed file of regions
    for which the methylation profile will be produced.

    {style("DATA_DIR", fg="green")} is the directory containing the methylation matrices
    produced by running 'scbs prepare'.

    {style("OUTPUT", fg="green")} is the file path where the methylation profile data
    will be written. Should end with '.csv'.
    """,
    short_help="Plot mean methylation around a group of genomic features",
    no_args_is_help=True,
)
@click.argument("regions", type=click.File("r"))
@click.argument(
    "data-dir",
    type=click.Path(exists=True, dir_okay=True, file_okay=False, readable=True),
)
@click.argument("output", type=click.File("w"))
@click.option(
    "--width",
    default=4000,
    show_default=True,
    type=click.IntRange(min=1, max=None),
    metavar="INTEGER",
    help="The total width of the profile plot in bp. "
    "The center of all bed regions will be "
    "extended in both directions by half of this amount. "
    "Shorter regions will be extended, longer regions "
    "will be shortened accordingly.",
)
@click.option(
    "--strand-column",
    type=click.IntRange(min=1, max=None),
    metavar="INTEGER",
    help="The bed column number (1-indexed) denoting "
    "the DNA strand of the region  [optional].",
)
@click.option(
    "--label",
    help="Specify a constant value to be added as a "
    "column to the output table. This can be "
    "useful to give each output a unique label when "
    "you want to concatenate multiple outputs  [optional].",
)
def profile_cli(**kwargs):
    timer = Timer(label="profile")
    _print_kwargs(kwargs)
    profile(**kwargs)
    timer.stop()


# smooth command
@cli.command(
    name="smooth",
    help=f"""
    This script will calculate the smoothed mean methylation over the
    whole genome.

    {style("DATA_DIR", fg="green")} is the directory containing the methylation matrices
    produced by running 'scbs prepare'.

    The smoothed methylation values will be written to
    {style("DATA_DIR/smoothed/", fg="green")}.
    """,
    short_help="Smooth the pseudobulk of single cell methylation data",
    no_args_is_help=True,
)
@click.argument(
    "data-dir",
    type=click.Path(
        exists=True, dir_okay=True, file_okay=False, readable=True, writable=True
    ),
)
@click.option(
    "-bw",
    "--bandwidth",
    default=1000,
    type=click.IntRange(min=1, max=1e6),
    metavar="INTEGER",
    show_default=True,
    help="Smoothing bandwidth in basepairs.",
)
@click.option(
    "--use-weights",
    is_flag=True,
    help="Use this to weigh each methylation site by log1p(coverage).",
)
def smooth_cli(**kwargs):
    timer = Timer(label="smooth")
    _print_kwargs(kwargs)
    smooth(**kwargs)
    timer.stop()


# matrix command (makes a "count" matrix)
@cli.command(
    name="matrix",
    help=f"""
    From single cell methylation or NOMe-seq data, calculates the average methylation
    in genomic regions for every cell. The output is a long table that can be used e.g.
    for dimensionality reduction or clustering, analogous to a count matrix in
    scRNA-seq.

    {style("REGIONS", fg="green")} is an alphabetically sorted (!) .bed file of regions
    for which methylation will be quantified in every cell.

    {style("DATA_DIR", fg="green")} is the directory containing the methylation
    matrices produced by running 'scbs prepare', as well as the smoothed methylation
    values produced by running 'scbs smooth'.

    {style("OUTPUT", fg="green")} is the file path where the count table will be
    written. Should end with '.csv'. The table is in long format and missing values
    are omitted.
    """,
    short_help="Make a methylation matrix, similar to a count matrix in scRNA-seq",
    no_args_is_help=True,
)
@click.argument("regions", type=click.File("r"))
@click.argument(
    "data-dir",
    type=click.Path(exists=True, dir_okay=True, file_okay=False, readable=True),
)
@click.argument("output", type=click.File("w"))
@click.option(
    "--keep-other-columns",
    is_flag=True,
    help="Use this to keep any other columns that the input bed-file may contain.",
)
def matrix_cli(**kwargs):
    timer = Timer(label="matrix")
    _print_kwargs(kwargs)
    matrix(**kwargs)
    timer.stop()


# scan command
@cli.command(
    name="scan",
    help=f"""
    Scans the whole genome for regions of variable methylation. This works by sliding
    a window across the genome, calculating the variance of methylation per window,
    and selecting windows above a variance threshold.

    {style("DATA_DIR", fg="green")} is the directory containing the methylation
    matrices produced by running 'scbs prepare', as well as the smoothed methylation
    values produced by running 'scbs smooth'.

    {style("OUTPUT", fg="green")} is the path of the output file in '.bed' format,
    containing the variable windows that were found.
    """,
    short_help="Find regions with variable methylation (work in progress)",
    no_args_is_help=True,
)
@click.argument(
    "data-dir",
    type=click.Path(exists=True, dir_okay=True, file_okay=False, readable=True),
)
@click.argument("output", type=click.File("w"))
@click.option(
    "-bw",
    "--bandwidth",
    default=2000,
    type=click.IntRange(min=1, max=1e6),
    metavar="INTEGER",
    show_default=True,
    help="Bandwidth of the variance windows in basepairs.",
)
@click.option(
    "--stepsize",
    default=50,
    type=click.IntRange(min=1, max=1e6),
    metavar="INTEGER",
    show_default=True,
    help="Step size of the variance windows in basepairs.",
)
@click.option(
    "--var-threshold",
    default=0.05,
    show_default=True,
    help="The variance threshold. The default of 0.05 means that the top 5% most variable "
    "genomic bins will be merged and reported.",
)
@click.option(
    "--threads",
    default=-1,
    help="How many CPU threads to use in parallel.  [default: all available]",
    callback=_set_n_threads,
)
def scan_cli(**kwargs):
    timer = Timer(label="scan")
    _print_kwargs(kwargs)
    scan(**kwargs)
    timer.stop()


# CLI template:
# @click.command(
#     help=f"""
#     Blabla

#     {style("INPUT", fg="green")} blabla

#     {style("OUTPUT", fg="green")} blabla
#     """,
#     short_help="template for dev",
# )
# @click.argument("input", type=click.File("r"))
# @click.argument("output", type=click.File("w"))
# @click.option("-o", "--option", type=int, default=4, show_default=True)
# @click.option("--flag", is_flag=True)
# def template(**kwargs):
#     timer = Timer(label="template")
#     _print_kwargs(kwargs)
#     print(**kwargs)
#     timer.stop()


# cli.add_command(matrix_cli, name="matrix")
# cli.add_command(profile_cli, name="profile")
